src/inmanta/util/dict_path.py

```diff
@@ -122,6 +122,15 @@ def escape(self) -> str:
     def matches(self, value: Optional[object]) -> bool:
         if value is None:
             return False
+
+        # Attempt numeric comparison if both values can be converted to floats
+        try:
+            if float(self._value) == float(value):
+                return True
+        except ValueError:
+            pass
+
+        # Fallback to string comparison
```

Comment:
This implementation will always attempt a numeric comparison, also when the provided value has the type `str`. I don't think that makes much sense. Wouldn't it be better to only attempt a numeric comparison when the provided value is numeric?

For example, consider the following dict: `{"1": "val"}`. If we apply the dict path expression `1.0` on this dict, do we want it to match? I personally think not. But I do think the dict path expression should match `{1: "val"}`. Do you agree?

Reply:
The values in the dictpaths are always provided as strings aren't they? I changed the code and added some test. I think it is better now.

Reply:
Yes. In a dichtpath expression we don't have a way to indicate the difference between a string and an integer/float key.